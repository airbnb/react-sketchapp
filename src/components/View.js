/* @flow */
import React from 'react';
import PropTypes from 'prop-types';
import StyleSheet from '../stylesheet';
import ViewStylePropTypes from './ViewStylePropTypes';
import ResizingConstraintPropTypes from './ResizingConstraintPropTypes';

const propTypes = {
  // TODO(lmr): do some nice warning stuff like RN does
  style: PropTypes.oneOfType([
    PropTypes.shape({ ...ViewStylePropTypes }),
    PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.shape({ ...ViewStylePropTypes }),
        PropTypes.number,
      ])
    ),
    PropTypes.number,
  ]),
  name: PropTypes.string,
  resizingConstraint: PropTypes.shape({
    ...ResizingConstraintPropTypes,
  }),
  children: PropTypes.node,
};

class View extends React.Component {
  static defaultProps = {
    name: 'View',
  };

  render() {
    return (
      <view
        name={this.props.name}
        style={StyleSheet.flatten(this.props.style)}
        resizingConstraint={this.props.resizingConstraint}
      >
        {this.props.children}
      </view>
    );
  }
}

View.propTypes = propTypes;

module.exports = View;
