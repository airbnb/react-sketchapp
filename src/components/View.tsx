import * as React from 'react';
import * as PropTypes from 'prop-types';
import { or } from 'airbnb-prop-types';
import StyleSheet from '../stylesheet';
import ViewStylePropTypes from './ViewStylePropTypes';
import ResizingConstraintPropTypes from './ResizingConstraintPropTypes';
import ShadowsPropTypes from './ShadowsPropTypes';

export const ViewPropTypes = {
  // TODO(lmr): do some nice warning stuff like RN does
  style: or([PropTypes.shape(ViewStylePropTypes), PropTypes.number]),
  name: PropTypes.string,
  resizingConstraint: PropTypes.shape({
    ...ResizingConstraintPropTypes,
  }),
  shadows: PropTypes.arrayOf(
    PropTypes.shape({
      ...ShadowsPropTypes,
    }),
  ),
  flow: PropTypes.shape({
    targetId: PropTypes.string,
    target: PropTypes.string,
    animationType: PropTypes.string,
  }),
  children: PropTypes.node,
};

export type Props = PropTypes.InferProps<typeof ViewPropTypes>;

export default class View extends React.Component<Props> {
  static propTypes = ViewPropTypes;

  static defaultProps = {
    name: 'View',
  };

  render() {
    return (
      <sketch_view
        name={this.props.name}
        style={StyleSheet.flatten(this.props.style)}
        resizingConstraint={this.props.resizingConstraint}
        shadows={this.props.shadows}
        flow={this.props.flow}
      >
        {this.props.children}
      </sketch_view>
    );
  }
}
