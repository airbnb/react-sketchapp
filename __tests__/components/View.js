import React from 'react';
import renderer from 'react-test-renderer';
import View from '../../src/components/View';
import StyleSheet from '../../src/stylesheet';

describe('<View />', () => {
  it('passes its children', () => {
    const tree = renderer.create(<View>foo</View>).toJSON();

    expect(tree).toMatchSnapshot();
  });

  describe('name', () => {
    it('passes its name', () => {
      const tree = renderer.create(<View name="foo" />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('defaults to View', () => {
      const tree = renderer.create(<View />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe('style', () => {
    const styles = StyleSheet.create({
      view: {
        flex: 1,
      },
    });

    it('accepts a plain object', () => {
      const tree = renderer.create(<View style={{ flex: 1 }} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts a StyleSheet ordinal', () => {
      const tree = renderer.create(<View style={styles.view} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('accepts an array of plain objects and/or StyleSheet ordinals', () => {
      const tree = renderer
        .create(<View style={[{ flexGrow: 1 }, styles.view]} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
