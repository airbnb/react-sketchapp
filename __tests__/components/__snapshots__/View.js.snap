exports[`<View /> name defaults to View 1`] = `
<view
  name="View"
  resizingConstraint={undefined}
  style={undefined} />
`;

exports[`<View /> name passes its name 1`] = `
<view
  name="foo"
  resizingConstraint={undefined}
  style={undefined} />
`;

exports[`<View /> passes its children 1`] = `
<view
  name="View"
  resizingConstraint={undefined}
  style={undefined}>
  foo
</view>
`;
